Realm/RLMSyncSession.h

```diff
@@ -30,11 +30,57 @@ typedef NS_ENUM(NSUInteger, RLMSyncSessionState) {
     RLMSyncSessionStateInvalid
 };
 
+/**
+ The type of data transfer that a particular sync session notifier block will track.
+ */
+typedef NS_ENUM(NSUInteger, RLMSyncNotifierDirection) {
+    /// For monitoring upload progress.
+    RLMSyncNotifierDirectionUpload,
+    /// For monitoring download progress.
+    RLMSyncNotifierDirectionDownload,
+};
+
+/**
+ The desired behavior of a particular sync session notifier block.
+ */
+typedef NS_ENUM(NSUInteger, RLMSyncNotifierMode) {
+    /// The block will be called forever, or until the user unregisters it.
+    /// It will always report the latest number of transferred bytes, and the
+    /// latest number of total bytes to be transferred.
+    RLMSyncNotifierModeAlwaysReportLatest,
+    /// The block will, upon registration, store the total number of bytes
+    /// to be transferred. When invoked, it will always report the latest number
+    /// of transferred bytes out of that original number of bytes to be
+    /// transferred. When the number of transferred bytes reaches or exceeds the
+    /// number of bytes to be transferred, the block will be unregistered.
+    RLMSyncNotifierModeProgressIndicator,
+};
+
 @class RLMSyncUser, RLMSyncConfiguration;
 
+/**
+ The type of a block intended for reporting a session's network activity to the user.
+
+ `transferred_bytes` refers to the number of bytes that have been uploaded or downloaded.
+
+ `transferrable_bytes` refers to the number of total bytes to be uploaded or downloaded.
+ */
+typedef void(^RLMProgressNotificationBlock)(NSUInteger transferred_bytes, NSUInteger transferrable_bytes);
+
 NS_ASSUME_NONNULL_BEGIN
 
 /**
+ A token object corresponding to a progress notifier block on a `RLMSyncSession`. To stop notifications manually,
+ destroy the token or call `-stop` on it.
+ */
+@interface RLMProgressNotificationToken : NSObject
```

Comment:
I don't think there's really a big gain there (IMO there's nothing unclear about the semantics of passing a sync progress notification cancelation token to beginWrite, even if it's pointless), and I think having a single cancelation token type used for absolutely everything makes our API more approachable. It's nice to have the property of that a user never has to think about what sort of token this notification method will return or what they can do with it.